### ISSUE TITLE
MongodbSerializer fails when entity id is assigned a String and is not an ObjectId

### ISSUE THREAD

**Issue by gaspardpetit** [querydsl/querydsl#1795]
This was introduced in the following changelist https://github.com/querydsl/querydsl/commit/3d09b9581bf1ef2c52118364c372ce640ec7f844

The following code expects entity Ids to be object Ids:

```java
protected Object convert(Path<?> property, Constant<?> constant) {
	        if (isReference(property)) {
	            return asReference(constant.getConstant());
	        } else if (isId(property)) {
	            if (isReference(property.getMetadata().getParent())) {
	                return asReferenceKey(property.getMetadata().getParent().getType(), constant.getConstant());
	            } else if (constant.getType().equals(String.class)) {
	                return new ObjectId((String) constant.getConstant());
	            }
	        }
	        return visit(constant, null);
	    }
```

Here is a simple unit test that fails:

```java
@Entity
public class WithStringId {
	@Id	String id;
}
```

```java
	@Test
	public void test_QueryDSL_StringIds()
	{
		IPrimaryMongoConnector mongo = ClassContainer.getInstance(IPrimaryMongoConnector.class);

		Morphia morphia = new Morphia();
		Datastore ds = morphia.createDatastore(mongo.getMongo(), "sampledb"); 
		MorphiaSerializer ser = new MorphiaSerializer(new Morphia());
		
		MorphiaQuery<WithStringId> query = new MorphiaQuery<>(morphia, ds, WithStringId.class);
		query = query.where(QWithStringId.withStringId.id.eq("a string value"));
		query.toString();
	}
```

An exception is thrown in the constructor of ObjectId, called in the MorphiaSerializer on the following line:

```java
                return new ObjectId((String) constant.getConstant());
```
In this case, the code seems to be a special case for when a String expression is provided for an ObjectId instead of an actual ObjectId.  The problem is that it does not ensure that the property type is an ObjectId.  A check is missing to make sure that property.type is indeed an ObjectId.

I would suggest the following change but I haven't tested:

```java
            } else if (constant.getType().equals(String.class)) {
```

should be

```java
            } else if (property.getType.equals(ObjectId.class) && constant.getType().equals(String.class)) {
```

Gaspard

**Comment by gaspardpetit** [comment]
I tested the fix - for some reason I am not able to create a pull request so here is the patch:

Date: Mon, 15 Feb 2016 20:04:41 -0500
Subject: [PATCH] Relates to issue 1795
 (https://github.com/querydsl/querydsl/issues/1795) : MongodbSerializer fails
 when entity id is assigned a String and is not an ObjectId The fix checks
 that the property is an ObjectId before attempting to convert a String value
 to ObjectId.  This is for a special case where the serializer accepts a
 String value for an ObjectId property.

---
 .../src/main/java/com/querydsl/mongodb/MongodbSerializer.java          | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

```
diff --git a/querydsl-mongodb/src/main/java/com/querydsl/mongodb/MongodbSerializer.java b/querydsl-mongodb/src/main/java/com/querydsl/mongodb/MongodbSerializer.java
index 578fcc3..bc242a1 100644
--- a/querydsl-mongodb/src/main/java/com/querydsl/mongodb/MongodbSerializer.java
+++ b/querydsl-mongodb/src/main/java/com/querydsl/mongodb/MongodbSerializer.java
@@ -305,7 +305,8 @@ public abstract class MongodbSerializer implements Visitor<Object, Void> {
         } else if (isId(property)) {
             if (isReference(property.getMetadata().getParent())) {
                 return asReferenceKey(property.getMetadata().getParent().getType(), constant.getConstant());
-            } else if (constant.getType().equals(String.class)) {
+            } else if (property.getType().equals(ObjectId.class) 
+            		&& constant.getType().equals(String.class)) {
                 return new ObjectId((String) constant.getConstant());
             }
         }
-- 
2.6.2
```

**Comment by timowest** [comment]
I am not sure if the fix attempt is valid, but I agree that the current behaviour is not correct.

**Comment by timowest** [comment]
Because of backwards compatibility it's problematic to drop the implicit conversion fully, but now it should be applied only when the ObjectId conversion is possible.

**Comment by gaspardpetit** [comment]
Hi,  I tested the change you made and unfortunately I don't think it is what we really want. 

I used the following entity:
```java
@Entity public class WithStringId {
	@Id String id;
}
```

With the following code:

```java
MorphiaQuery<WithStringId> query = new MorphiaQuery<>(morphia, ds, WithStringId.class);
query = query.where(QWithStringId.withStringId.id.eq("56c7f5bb4f636732e414b43d"));
String value = query.toString();
```

The following asserts:

```java
Assert.assertEquals("{ \"_id\" : \"56c7f5bb4f636732e414b43d\"}", value);
```

Instead of ```{ "_id" : "56c7f5bb4f636732e414b43d"}```, I obtain ```{ "_id" : { "$oid" : "56c7f5bb4f636732e414b43d"}}```

In this case, I am assigning a ``String`` to a ``String`` - that ``String`` happens to be parseable as an ``ObjectId``, but with the resulting query, mongo will search for an _id of type ``ObjectId`` instead of searching for an _id of type ``String``.  This is dangerous since it will not match a document whose id is a string with value "56c7f5bb4f636732e414b43d" and this will cause bugs that will appear at random (when the string happens to be in the format of an ``ObjectId`` - whether intended or by luck.

The fix that I suggested above did not remove the implicit conversion, it simply makes sure that the type to which we are assigning a value is an ``ObjectId``.  If it is, the implicit conversion will occur.


Gaspard
